Added a race condition test

Orbittman opened the pull request:
This is to add an extra test around the race condition bug that I had recently. I thought that it would give a good idea about how someone would implement a threaded solution. The're is nothing in the readme yet, this I will add later and the NCrunch files I'll remove. Don't bother committing this but some review comments would be welcome. There is also some refactoring of separate files that seems to have not been included, I'll refactor that at some point before we push this to master.

GrahamClark commented on UnitTests/RaceConditionTests.cs:
```diff
@@ -0,0 +1,52 @@
+                    .When(x => x.Set(Arg.Any<ISession>()))
+                    .Do(s =>
+                    {
+                        Thread.Sleep(latency);
```
should we be using `await Task.Delay(latency)` instead? Does this still cause the issue?

Orbittman replied:
It doesn't work as it's just there to simulate a pause not to change the flow of the code. Changing this breaks the simulated action.

GrahamClark commented on UnitTests/RaceConditionTests.cs:
```diff
@@ -0,0 +1,52 @@
+            foreach (var session in sessions)
+            {
+                sessionManager.Get(session.Key)
+                    .Returns(k => new Session { Key = (string)k[0], Count = sessions.Single(s => s.Key == (string)k[0]).Count });
```
in the NSubstitute lambdas like this, the parameter is of type `CallInfo`, which is an array of `object`s representing the parameters passed to the mocked method. So for clarity I think it's better to call these lambda parameters something like `info`, i.e. `.Returns(info => new Session...`

GrahamClark commented on UnitTests/RaceConditionTests.cs:
```diff
@@ -0,0 +1,52 @@
+                    .Do(s =>
+                    {
+                        Thread.Sleep(latency);
+                        sessions.Single(t => t.Key == ((ISession)s[0]).Key).Count = ((ISession)s[0]).Count;
```
same here, the lambda parameters could be a bit clearer as `s` kind of implies it's a Session object or similar, when actually it's just an object array wrapper.

Orbittman commented:
@GrahamClark @AnturGyffrous 
I've updated the test so that all of session request run in parallel, previously they were running synchronously. The solution that we have in our code base works with this so I know the test is valid and passable. I've updated the readme to include instructions for the new test, if there are no objections I will merge this in and pass it to the .NET contractors that we are recruiting.

AnturGyffrous commented:
Just a thought, when it get's merged in our comments will be visible
publicly. Would it be better to delete the pull request and branch and just
re-add the files as they are now?

GrahamClark commented:
@AnturGyffrous is that because you're worried people will spot your strange use of apostrophes? :wink:

AnturGyffrous commented:
Wh'ats wro'ng wit'h my a'postroph'es?

GrahamClark commented:
I don't have a problem with this being public.
